### PR TITLE
PLM-3|Jithya| API to update user details

### PR DESCRIPTION
**Pull request by jithyamukund**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -89,6 +90,7 @@ group :test do
   gem "selenium-webdriver"
   gem "webdrivers", '= 5.3.0'
   gem 'rspec-rails', '~> 6.1.0'
+  gem 'rails-controller-testing'
```
Comment (vijithmv499) [inline Gemfile:93]: why is this required?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,32 @@
...
+  def authenticate_request
+    header = request.headers['Authorization']
+    token = header&.split&.last
+
+    return render json: { error: 'Unauthorized' }, status: :unauthorized unless token
+
+    begin
+      decoded_token = decode_token(token)
+    rescue JWT::ExpiredSignature
+      return render json: { error: 'Token has expired' }, status: :unauthorized
```
Comment (vijithmv499) [inline app/controllers/application_controller.rb]: Now that you have introduced I18n localization, can we move these hardcoded errors to the en.yaml file?

Comment (vijithmv499) [reply app/controllers/application_controller.rb]: This applies to all such use-cases

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,32 @@
...
+    rescue JWT::DecodeError
+      return render json: { error: 'Unauthorized' }, status: :unauthorized
+    end
+
+    @current_user = User.find_by(mobile_number: decoded_token[:mobile_number])
+    render json: { error: 'Unauthorized' }, status: :unauthorized unless @current_user
+
+    # refresh token flow
```
Comment (vijithmv499) [inline app/controllers/application_controller.rb:26]: Please make sure a ticket is created for the same

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,32 @@
...
+  end
+
+  def decode_token(token)
+    JWT.decode(token, Rails.application.credentials.jwt_private_key, true)[0].with_indifferent_access
+  end
+
+  attr_reader :current_user
```
Comment (vijithmv499) [inline app/controllers/application_controller.rb]: Isn't it more readable to keep this attr_reader on the top of the class definition? Lets have a discussion if you have any counter arguments to support this way of declaration

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      render json: { user: user }, status: :ok
+    else
+      render json: { errors: [user.errors.full_messages.join(', ')] }, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def users_params
+    params.require(:user).permit(:first_name, :last_name, :email)
```
Comment (vijithmv499) [inline app/controllers/users_controller.rb:18]: 👍

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+class UsersController < ApplicationController
+  def update
+    user = User.find_by(id: params[:id])
+    return render json: { errors: ['User not found'] }, status: :not_found unless user
+
+    if user.update(users_params)
+      render json: { user: user }, status: :ok
```
Comment (vijithmv499) [inline app/controllers/users_controller.rb]: I've been adding the success response inside a `data` attribute for the previously implemented APIs? Can we have the same pattern here as well to make it consistent with other APIs? I'm okay to have a discussion and change the other API response formats if you have any valid concerns around this response structure. Let me know

File: app/services/authentication_service.rb [hunk 7]
```diff
@@ -35,7 +35,7 @@ def process
 
   def generate_token_response(user)
     token = JWT.encode({ mobile_number: user.mobile_number, exp: OTP_EXPIRY_IN_HOURS.hours.from_now.to_i },
-                       Rails.application.secrets.jwt_private_key)
+                       Rails.application.credentials.jwt_private_key)
```
Comment (vijithmv499) [inline app/services/authentication_service.rb:38]: 👍

File: config/locales/en.yml [hunk 8]
```diff
@@ -31,3 +31,4 @@
 
 en:
   hello: "Hello world"
+  is_invalid: 'is Invalid'
```
Comment (vijithmv499) [inline config/locales/en.yml]: can we move this inside `errors` key?

File: app/models/user.rb [hunk 9]
```diff
@@ -1,2 +1,3 @@
 class User < ApplicationRecord
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP, message: I18n.t('is_invalid') }, on: :update
```
Comment (vijithmv499) [inline app/models/user.rb]: Please add a spec for this validation

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+class UsersController < ApplicationController
+  def update
+    user = User.find_by(id: params[:id])
+    return render json: { errors: ['User not found'] }, status: :not_found unless user
+
+    if user.update(users_params)
```
Comment (Navinnair) [inline app/controllers/users_controller.rb]: Given that we've authenticated the request and established `current_user`, would it be appropriate to use `current_user.update`? Please let me know if I'm overlooking any considerations.

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+      render json: { user: user }, status: :ok
+    else
+      render json: { errors: [user.errors.full_messages.join(', ')] }, status: :unprocessable_entity
```
Comment (Navinnair) [inline app/controllers/users_controller.rb]: Have you looked into Rails' built-in `to_sentence` method? We could potentially replace `user.errors.full_messages.join(', ') `with `user.errors.full_messages.to_sentence`

File: app/models/user.rb (hunk 9 above)
Comment (Navinnair) [inline app/models/user.rb]: Have we considered validating the email format whenever there's an email input, rather than only on update? We could potentially remove the on: :update option to ensure consistent email validation across all scenarios where the email might be set or changed. 

something like

`validates :email, format: { with: URI::MailTo::EMAIL_REGEXP, message: I18n.t('is_invalid') }, if: -> { email.present?  }`

File: config/routes.rb [hunk 12]
```diff
@@ -1,9 +1,11 @@
 Rails.application.routes.draw do
+  resources :parking_lots
```
Comment (Navinnair) [inline config/routes.rb]: :thinking:

**Review by vijithmv499 — APPROVED** [review]

**Review by Navinnair — APPROVED** [review]
LGTM :rocket:
